src/libsyntax/feature_gate.rs

```diff
@@ -1220,72 +1221,114 @@ pub const BUILTIN_ATTRIBUTES: &[(&str, AttributeType, AttributeTemplate, Attribu
-                               "optimize_attribute",
-                               "#[optimize] attribute is an unstable feature",
-                               cfg_fn!(optimize_attribute))),
+    (optimize, Whitelisted, template!(List: "size|speed"), Gated(Stability::Unstable,
+                            "optimize_attribute",
+                            "#[optimize] attribute is an unstable feature",
+                            cfg_fn!(optimize_attribute))),
 
     // Crate level attributes
-    ("crate_name", CrateLevel, template!(NameValueStr: "name"), Ungated),
-    ("crate_type", CrateLevel, template!(NameValueStr: "bin|lib|..."), Ungated),
-    ("crate_id", CrateLevel, template!(NameValueStr: "ignored"), Ungated),
-    ("feature", CrateLevel, template!(List: "name1, name1, ..."), Ungated),
-    ("no_start", CrateLevel, template!(Word), Ungated),
-    ("no_main", CrateLevel, template!(Word), Ungated),
-    ("no_builtins", CrateLevel, template!(Word), Ungated),
-    ("recursion_limit", CrateLevel, template!(NameValueStr: "N"), Ungated),
-    ("type_length_limit", CrateLevel, template!(NameValueStr: "N"), Ungated),
-    ("test_runner", CrateLevel, template!(List: "path"), Gated(Stability::Unstable,
+    (crate_name, CrateLevel, template!(NameValueStr: "name"), Ungated),
+    (crate_type, CrateLevel, template!(NameValueStr: "bin|lib|..."), Ungated),
+    (crate_id, CrateLevel, template!(NameValueStr: "ignored"), Ungated),
+    (feature, CrateLevel, template!(List: "name1, name1, ..."), Ungated),
+    (no_start, CrateLevel, template!(Word), Ungated),
+    (no_main, CrateLevel, template!(Word), Ungated),
+    // FIXME: Duplicated
+    //(no_builtins, CrateLevel, template!(Word), Ungated),
+    (recursion_limit, CrateLevel, template!(NameValueStr: "N"), Ungated),
+    (type_length_limit, CrateLevel, template!(NameValueStr: "N"), Ungated),
+    (test_runner, CrateLevel, template!(List: "path"), Gated(Stability::Unstable,
                     "custom_test_frameworks",
                     EXPLAIN_CUSTOM_TEST_FRAMEWORKS,
                     cfg_fn!(custom_test_frameworks))),
-];
+}}}
+
+type BuiltinAttribute = (Symbol, AttributeType, AttributeTemplate, AttributeGate);
+
+macro_rules! declare_builtin_attributes {
+    ($(($name:ident, $level:expr, $template: expr, $gate: expr),)*) => {
+        /// Attributes that have a special meaning to rustc or rustdoc
+        pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
+            $((symbols::$name, $level, $template, $gate)),*
+        ];
+    }
+}
+
+macro_rules! find_builtin_attribute_int {
+    ($(($num:expr, ($name:ident, $level:expr, $template: expr, $gate: expr)))*) => {
+        pub fn find_builtin_attr(name: Symbol) -> Option<&'static BuiltinAttribute> {
+            // Check for duplicate attributes
+            $(#[allow(unused_imports)] use symbols::$name;)*
+
+            debug_assert!(!name.is_gensymed());
```

Comment:
It's ok to have a gensym here or in `impl PartialEq<Symbol> for Path`, we should just compare symbols (thus treating gensyms as different/not equal to non-gensym idents).